src/GHC/RTS/Events/Sort.hs

```diff
@@ -0,0 +1,188 @@
+-- | Constant-space sorting.
+--
+-- This module provides a routine for sorting events in constant-space via
+-- on-disk merge sort.
+module GHC.RTS.Events.Sort
+  ( sortEvents
+  , sortEvents'
+  , SortParams(..)
+  , defaultSortParams
+  ) where
+
+import Data.Traversable
+import Data.Coerce
+import Data.Function (on)
+import Data.List (sortBy, minimumBy)
+import Data.Maybe
+import Data.Foldable (toList)
+import System.IO
+import System.IO.Temp
+import System.Directory
+import Prelude
+
+import Data.Binary.Put as P
+import qualified Data.ByteString.Lazy as BSL
+import qualified Data.Sequence as S
+
+import GHC.RTS.Events hiding (sortEvents)
+import GHC.RTS.Events.Binary (putEventLog)
+
+type SortedChunk = FilePath
+
+newtype OnTime = OnTime Event
+
+instance Ord OnTime where
+  compare = coerce (compare `on` evTime)
+
+instance Eq OnTime where
+  (==) = coerce ((==) `on` evTime)
+
+-- | Parameters which determine the behavior of the merge sort.
+data SortParams = SortParams
+  { -- | The chunk size which the input eventlog is broken into (in events). This
+    -- determines the upper-bound on memory usage during the sorting process.
+    --
+    -- This value is a reasonable trade-off between memory and computation,
+    -- requiring approximately 100MBytes while sorting a "typical" eventlog.
+    chunkSize :: !Int
+
+    -- | Maximum number of chunks to merge at once. Determined by the largest
+    -- number of file descriptors we can safely open at once.
+  , maxFanIn  :: !Int
+  }
+
+-- | A reasonable set of sorting parameters.
+defaultSortParams :: SortParams
+defaultSortParams =
+  SortParams { chunkSize = 500*1000
+             , maxFanIn  = 256
+             }
+
+-- | @sortEvents outPath eventlog@ sorts @eventlog@ via on-disk merge
+-- sort. The sorted eventlog is written to @eventlog@. The system's temporary
+-- directory is used for temporary data. See 'sortEvents\'' for more control.
+sortEvents
+  :: FilePath  -- ^ output eventlog file path
+  -> EventLog  -- ^ eventlog to sort
+  -> IO ()
+sortEvents outPath eventLog =
+  withSystemTempDirectory "sort-events" $ \tmpDir ->
+    sortEvents' defaultSortParams tmpDir outPath eventLog
+
+-- | @sortEvents' params tmpDir outPath eventlog@ sorts
+-- @eventlog@ via on-disk merge sort, using @tmpDir@ for
+-- intermediate data. The caller is responsible for deleting @tmpDir@ upon
+-- return.
+--
+-- The sorted eventlog is written to @eventlog@.
+sortEvents'
+  :: SortParams
+  -> FilePath  -- ^ temporary directory
+  -> FilePath  -- ^ output eventlog file path
+  -> EventLog  -- ^ eventlog to sort
+  -> IO ()
+sortEvents' _params _tmpDir _outPath (EventLog _ (Data [])) = fail "sortEvents: no events"
```

Comment:
I think this function doesn't have to fail on an empty eventlog. Can we just `return ()`?

Reply:
This would be surprising as the output path wouldn't exist. However, I suppose we could just write out an empty eventlog.

Reply:
Hmm I see your point. Yeah the current behavior is preferable I guess.